### PR TITLE
[Refactor] Extract redux base classes to common module

### PR DESCRIPTION
**Pull request by emlynmac**
## Purpose
Bring out the redux core interfaces to the common code

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

### CODE REVIEW

File: AzureCommunicationUI/.swiftlint.yml [hunk 1]
```diff
@@ -165,7 +165,7 @@ opt_in_rules:
   - weak_delegate
 
 disabled_rules: # rule identifiers turned on by default to exclude from running
-
+ - function_parameter_count
```
Comment (JoshuaLai) [inline AzureCommunicationUI/.swiftlint.yml:168]: Why did we want to disable this rule? Was there a method that was triggering it?

File: AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Redux/StoreExtensions.swift [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            reducer: .appStateReducer(),
+            middlewares: [
+                .liveCallingMiddleware(
+                    callingMiddlewareHandler: CallingMiddlewareHandler(
+                        callingService: callingService,
+                        logger: logger
+                    )
+                )
+            ],
+            state: AppState(localUserState: localUserState)
```
Comment (JoshuaLai) [inline AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Redux/StoreExtensions.swift:29]: Perhaps a follow up task and not for this PR. 
If we are going to Explicitly call out `ChatAppState` this AppState should be `CallingAppState`

Comment (emlynmac) [reply AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Redux/StoreExtensions.swift:29]: Yes, for sure.

File: AzureCommunicationUI/.swiftlint.yml (hunk 1 above)
Comment (emlynmac) [reply AzureCommunicationUI/.swiftlint.yml:168]: Yes. The construct store method.

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatActionHandler.swift [hunk 3]
```diff
@@ -23,35 +23,36 @@ protocol ChatActionHandling {
...
+                       state: ChatAppState,
                        dispatch: @escaping ActionDispatch) -> Task<Void, Never>
     @discardableResult
-    func sendTypingIndicator(state: AppState, dispatch: @escaping ActionDispatch) -> Task<Void, Never>
+    func sendTypingIndicator(state: ChatAppState, dispatch: @escaping ActionDispatch) -> Task<Void, Never>
     @discardableResult
-    func sendReadReceipt(messageId: String, state: AppState, dispatch: @escaping ActionDispatch) -> Task<Void, Never>
-    func setTypingParticipantTimer(_ getState: @escaping () -> AppState, _ dispatch: @escaping ActionDispatch)
+    func sendReadReceipt(messageId: String, state: ChatAppState, dispatch: @escaping ActionDispatch)
+    -> Task<Void, Never>
```
Comment (JoshuaLai) [inline AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatActionHandler.swift:54]: Assuming this was cause swift lint was throwing a character length violation

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatMiddleware.swift [hunk 4]
```diff
@@ -36,89 +36,89 @@ extension Middleware {
...
     }
-}
 
-private func handleLifecycleAction(_ action: LifecycleAction,
-                                   _ actionHandler: ChatActionHandling,
-                                   _ getState: () -> AppState,
-                                   _ dispatch: @escaping ActionDispatch) {
-    print("`handleLifecycleAction` not implemented")
-}
+    private static func handleLifecycleAction(_ action: LifecycleAction,
```
Comment (JoshuaLai) [inline AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatMiddleware.swift:40]: Is there a reason to declare these methods as static?

**Review by JoshuaLai — COMMENTED** [review]
Tested and everything looks fine. Just a few questions behind the change in the middleware to switch to using static methods

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatActionHandler.swift (hunk 3 above)
Comment (emlynmac) [reply AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatActionHandler.swift:54]: Yup

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatMiddleware.swift (hunk 4 above)
Comment (emlynmac) [reply AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Redux/Middleware/ChatMiddleware.swift:40]: They need to be called from a static context. 
They were previously private file functions, but I need them in the class to make the generics work properly

**Review by JoshuaLai — DISMISSED** [review]

File: AzureCommunicationUI/.swiftlint.yml (hunk 1 above)
Comment (vhuseinova-msft) [inline AzureCommunicationUI/.swiftlint.yml:168]: Should we just setup warnings/error rules for` function_parameter_count` instead of disabling it fully?

**Review by vhuseinova-msft — DISMISSED** [review]

**Review by jimchou-dev — DISMISSED** [review]

File: AzureCommunicationUI/.swiftlint.yml (hunk 1 above)
Comment (emlynmac) [reply AzureCommunicationUI/.swiftlint.yml]: For now, I've relaxed it a bit but kept it enabled.
See https://realm.github.io/SwiftLint/function_parameter_count.html

**Review by vhuseinova-msft — APPROVED** [review]

**Review by JoshuaLai — APPROVED** [review]

**Pull request merged**